### PR TITLE
Migrates remaining warp-scope tests to Catch2

### PR DESCRIPTION
**Pull request by elstehle**
This PR ports the following warp-scope tests to Catch2:
- [x] `test/test_warp_reduce.cu`
- [x] `test/test_warp_merge_sort.cu`
- [x] `test/test_warp_load.cu`

### CODE REVIEW

File: test/catch2_test_warp_load.cu [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+         items_per_thread,
+         algorithm,
+         cache_load_modifier)
+{
+  using params                                   = params_t<TestType>;
+  using type                                     = typename params::type;
+  using delegate_t                               = unguarded_load_t;
+  constexpr cub::CacheLoadModifier load_modifier = c2h::get<4, TestType>::value;
+
+  const int valid_items =
```
Comment (senior-zero) [inline test/catch2_test_warp_load.cu:370]: Not referenced?

File: test/catch2_test_warp_load.cu [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+         types,
+         logical_warp_threads,
+         items_per_thread,
+         algorithm)
+{
+  using params     = params_t<TestType>;
+  using type       = typename params::type;
+  using delegate_t = unguarded_load_t;
+
+  const int valid_items =
```
Comment (senior-zero) [inline test/catch2_test_warp_load.cu:334]: Not referenced?

File: test/catch2_test_warp_merge_sort.cu [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+ */
+enum class stability
+{
+  stable,
+  unstable
+};
+
+// List of key types to test
+using custom_t  = c2h::custom_type_t<c2h::equal_comparable_t, c2h::less_comparable_t>;
+using key_types = c2h::type_list<std::uint8_t, std::int32_t, std::int64_t, custom_t>;
```
Comment (senior-zero) [inline test/catch2_test_warp_merge_sort.cu:390]: Do you think there might be a value in testing custom types support only for a subset of parameters? For instance, one warp size / items per thread?

File: test/catch2_test_warp_reduce.cu [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+
+// List of types to test
+using custom_t =
+  c2h::custom_type_t<c2h::accumulateable_t, c2h::equal_comparable_t, c2h::less_comparable_t>;
+using types = c2h::type_list<std::uint8_t,
+                             std::uint16_t,
+                             std::int32_t,
+                             std::int64_t,
+                             custom_t,
+                             ulonglong4,
```
Comment (senior-zero) [inline test/catch2_test_warp_reduce.cu:334]: Do you think there might be a value in extracting test of vector types out of the full configuration space?

**Review by senior-zero — APPROVED** [review]
A few minor changes. Once those are addressed, please, rebase and start run gpuCI.

File: test/c2h/custom_type.cuh [hunk 5]
```diff
@@ -72,8 +72,9 @@ class less_comparable_t
 public:
   __host__ __device__ bool operator<(const CustomType& other) const
   {
-    return static_cast<const CustomType&>(*this).get_key() 
-         < other.get_key();
+    return static_cast<const CustomType &>(*this).get_key() < other.get_key() ||
```
[the remark this answers is not in the document]
Comment (elstehle) [reply test/c2h/custom_type.cuh:90]: I've added `lexicographical_less_comparable_t` and making use of it for `reduce` (to make sure `Min()` works with `equal_comparable_t`) and `marge_sort`.

File: test/catch2_test_warp_load.cu [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+    // items/threads parameters are swapped
+
+    constexpr int FAKE_BLOCK_SIZE = ITEMS_PER_THREAD * TOTAL_WARPS;
+
+    fill_striped<ITEMS_PER_THREAD, LOGICAL_WARP_THREADS, FAKE_BLOCK_SIZE>(h_input.begin());
+    d_input = h_input;
+  }
+  else
+  {
+    thrust::sequence(d_input.begin(), d_input.end());
```
[the remark this answers is not in the document]
Comment (elstehle) [reply test/catch2_test_warp_load.cu]: Thanks! Adopted your suggestion.

File: test/catch2_test_warp_load.cu [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+
+  warp_load<params::algorithm,
+            params::logical_warp_threads,
+            params::items_per_thread,
+            params::total_warps,
+            type>(thrust::raw_pointer_cast(d_in.data()),
+                  delegate_t{valid_items, oob_default},
+                  thrust::raw_pointer_cast(d_error_counter.data()));
+
+  const auto num_errors          = d_error_counter[0];
```
[the remark this answers is not in the document]
Comment (elstehle) [reply test/catch2_test_warp_load.cu]: Agreed. Done 👍

**Pull request merged**